### DISCUSSION TITLE
ECC Encryption in WebCrypto is hard to implement

### DISCUSSION THREAD

**Discussion by Zhaoxun** [mdn/mdn-community#636] [Content] [answered]
Hi I am on a project to receive an ECDH pubkey from backend in order to encrypt messages so that only the true backend can decrypted, for I have to use a proxy site to pass clients' message to my actual server. I have checked pages like https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/generateKey & /importKey & /exportKey &etc.

First thing first, generate a public key:
```
curve = { name: 'ECDH', namedCurve: 'P-256' };
//k = await crypto.subtle.generateKey(curve, true, ['deriveKey']);
//pub = new Uint8Array( await crypto.subtle.exportKey('raw', k.publicKey) );
crypto.subtle.generateKey(curve, true, ["deriveKey"] ).
 then( key=>crypto.subtle.exportKey('raw', key.publicKey) ).
  then( raw=> pub = new Uint8Array(raw) )
```
I do not know why the `await` syntax worked in browser console but failed when embedded in a web page.
Anyway, since the public key `pub` is produced, let me try to import it:
```
crypto.subtle.importKey( "raw", pub, curve, "true",['deriveKey'] ).
  then(key => console.log(key))
```
Then there is a DOM Exception:`Cannot create a key using the specified key usages.` I cannot use `['encrypt']` either. So does it mean that encryption through an ECC public key is Not Supported?

**Comment by Zipdox2** [comment, marked as answer]
> So does it mean that encryption through an ECC public key is Not Supported?

Correct: https://w3c.github.io/webcrypto/#algorithm-overview

**Reply by bsmth** [reply]
Thanks, we also have a table here: https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto#supported_algorithms

**Comment by Zipdox2** [comment]
Even though encryption with ECC is not directly possible, you can take the PGP approach. I'm not gonna go into detail as I'm no expert on the matter, but basically you use an ephemeral key and ECDH with a KDF to derive a key.
